data/tools/python/metalistener/eiger_meta_writer.py

```diff
@@ -4,557 +4,144 @@
-
-        :param message: The message received
-        """
-        self._logger.debug('Handling global header none for acqID ' + self._acquisition_id)
-        self._logger.debug(message)
-
-        if self._series_created:
-            self._logger.debug('series already created')
-            return
-
-        if not self.file_created:
-            self.create_file()
-
-        npa = np.array(message['series'])
-        self.create_dataset_with_data("series", data=npa)
-        self._series_created = True
-
-        return
-
-    def handle_global_header_config(self, header, config):
-        """Handle global header config part message containing config data.
+import numpy as np
 
-        :param header: The header received
-        :param config: The config data
-        """
-        self._logger.debug('Handling global header cfg for acqID ' + self._acquisition_id)
-        self._logger.debug(header)
-        self._logger.debug(config)
+from odin_data.meta_writer.meta_writer import MetaWriter, MESSAGE_TYPE_ID, FRAME
+from odin_data.meta_writer.hdf5dataset import Int64HDF5Dataset, StringHDF5Dataset
+from odin_data.util import construct_version_dict
+import _version as versioneer
 
-        if not self.file_created:
-            self.create_file()
+# Data message parameters
+START_TIME = "start_time"
+STOP_TIME = "stop_time"
+REAL_TIME = "real_time"
+SERIES = "series"
+SIZE = "size"
+HASH = "hash"
+ENCODING = "encoding"
+DATATYPE = "type"
 
-        if self._config_created:
-            self._logger.debug('config already created')
-        else:
-            nps = np.str(config)
-            config_data = ast.literal_eval(np.str(config).decode("utf-8"))
-            self.create_dataset_with_data("config", data=nps)
-            for k in sorted(config_data):
-                self.create_dataset_with_data("_dectris/%s" %k, config_data[k])
-            self._config_created = True
 
-        if self._series_created:
-            self._logger.debug('series already created')
+class EigerMetaWriter(MetaWriter):
+    """Implementation of MetaWriter that also handles Eiger meta messages"""
+
+    # Define Eiger-specific datasets
+    DETECTOR_DATASETS = [
+        Int64HDF5Dataset(START_TIME),
+        Int64HDF5Dataset(STOP_TIME),
+        Int64HDF5Dataset(REAL_TIME),
+        Int64HDF5Dataset(SERIES),
+        Int64HDF5Dataset(SIZE),
+        StringHDF5Dataset(HASH, length=32),
+        StringHDF5Dataset(ENCODING, length=10),
+        StringHDF5Dataset(DATATYPE, length=6),
+    ]
+    # Define parameters received on per frame meta message
+    DETECTOR_WRITE_FRAME_PARAMETERS = [
+        START_TIME,
+        STOP_TIME,
+        REAL_TIME,
+        SERIES,
+        SIZE,
+        HASH,
+        ENCODING,
+        DATATYPE,
+    ]
+
+    def process_message(self, header, data):
+        """Process a message from a data socket
+
+        Check for Eiger-specific messages and pass anything else to base class
+
+        """
+        message_handlers = {
```

Comment:
Why is message handlers declared every time a message is processed?

Reply:
I think I tried to make it a class variable, but it didn't work because they need to be defined as `self.method` so I just put them in there. I guess they could go in `__init__`, though.

Reply:
Now that I think about it. It would be much cleaner to make a `message_handlers` property method and then also the rest of this method can go in the base class instead of being duplicated in each detector. What do you think?